### PR TITLE
Fixing attention masks in HF implementation

### PR DESCRIPTION
**Pull request by NohTow**
Hello, 

When trying to train the HF version of the model to do image captioning, I realized that the MLE loss was (_too_) quickly decreasing while not being able to generate proper captions and so achieve poor BLEU scores.

I digged into it and found that it comes from the attention masks. Indeed, when expanding the mask, tokens which correspond to a 1 in the input attention mask are replaced with -inf, when it should be these corresponding to 0 (1 mean that we are attending to this token for HF tokenizers). 

I fixed the line and also inverted the computed encoder masks (because it outputs 1 for padding tokens, the opposite of HF tokenizers).

Now it is learning properly, hope it helps.

PS: I don't know why/how, but I had a different version of modeling_ofa.py when I cloned the repo, and in this one, there was another mistake. In my version, I had this function: 
```
def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_values_length: int = 0):
    """
    Make causal mask used for bi-directional self-attention.
    """
    bsz, tgt_len = input_ids_shape
    mask = torch.full((tgt_len, tgt_len), 1)
    mask_cond = torch.arange(mask.size(-1))
    mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
    mask = mask.to(dtype)
    if past_key_values_length > 0:
        mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype), mask], dim=-1)
    mask = mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
    # return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
    return mask
```

Whereas, the second line should be `mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min)` (because masked token should be replaced with -inf so the softmax in the attention layer outputs 0 for them). Yet, I cannot find why I had this version, which is different from the one on the repo (even the documentation mention "bi-directional self-attention" whereas on the repository, it has always been "uni-directional" attention). This was why the MLE loss was decreasing so fast, the causal mask was wrong and so future tokens leaked.

### CODE REVIEW

**Comment by JustinLin610** [comment]
Thanks for your help! Will have a check about the details in these days:) Seemingly these are our bugs and your changes should help.

**Comment by JustinLin610** [comment]
Hey. Our logic is actually: 1. masks for tokens that should be attended are 0, and those for tokens that should not be attended are -inf. If you check the implementation of the attention module, we actually add the masks to the attention weights, so that those that should not be attended should be assigned with a very low score, whole the others keep unchanged.

**Comment by NohTow** [comment]
Hello,
Yes, values _added in the attention module_ should be 0 for attended tokens and -inf for the other, so the softmax actually mask unattended tokens.
The problem is that, when calculating the attention mask composed of 0 and 1 (just indicating which tokens are padding tokens after the tokenization), HF logic is to have 0 for padding/masked tokens and 1 for tokens we are attending to.
Then, this mask is translated by putting -inf for 0 values and 0 for 1 values.

So we agree on the implementation of the attention module, which need -inf for masked token and 0 for not masked tokens, but the original "attention_mask" produced by the tokenizer or so, should be 0 for masked tokens and 1 for not masked tokens. This is how it is defined in every HF tokenizer and modules. See : https://discuss.huggingface.co/t/why-does-bart-decoders-attention-mask-mark-relevant-indices-with-0-instead-of-1/6477/2

**Comment by NohTow** [comment]
Yes, this is mainly the problem in encountered.
I understand that it might seems like not a big deal (since you do not rely on this for your training), but I still think it might be a good thing to follow "HF logic" for the HF implementation. Especially since the OFATokenizer that is implemented, return this kind of mask. 
Besides this particular training, every scripts built for HF follow this logic, and I'm pretty sure it will be harmful in the future and even harder to detect that this is the cause.

But yeah, it makes sense that you did not face this issue before.

**Comment by JustinLin610** [comment]
Alright, got it! Codes are already merged. Actually I have no experience in training OFA with transformers. @faychu could share some experience, and in the near future, we might update the part of code.

**Comment by NohTow** [comment]
Very cool, thanks ! 
I am actively working on this rn, so if I can be of any help, do not hesitate to reach !
